arch: arm: cortex_m: Allow relocation of vector table in RAM

Martinhoff-maker opened the pull request:
**The goal of this PR is to add the support of vector table relocation for cortex_m architecture in case the symbol CONFIG_SRAM_VECTOR_TABLE is set to yes and VTOR register is present.**

Actually, as said in the current Kconfig.zephyr, the symbol CONFIG_SRAM_VECTOR_TABLE should place the vector table at the start of the SRAM (which indirectly place the irq_vector_table in the SRAM too).

CONFIG_SRAM_VECTOR_TABLE is currently only used by some boards/soc:

- Nuvoton/npcm400_evb -> activate by default but don't see any use of the symbol? (might add maxdog988 NUVOTON NPCM platform maintainer on the review)
- adi/max32 -> activate by default on all the max32 boards but no use of it as I see ? (might add MaureenHelm ADI maintener on the review) (edit : confirmation that this is not used by adi/max32)
- renesas/smartbond/da1469x -> It has an m33 then have a VTOR register. Actually, it's set to yes by default in the defconfig file of the soc. The symbol is not used at all, but they do unconditionnaly reserve some space at the start of the SRAM (might add ioannis-karachalios RENESAS smartbond maintainer on the review) (edit : PR to remove the symbol)
- st/stm32f0x -> they didn't have VTOR for the specific board that uses CONFIG_SRAM_VECTOR_TABLE (but it's where I take inspiration for this PR) so it should not enter into conflict. (might add erwango ST maintainer on the review) (edit : no problem, seen with mathieuchopstm)

It's hence not generic at all and would like to make this more generic for arm/cortex_m architecture which have VTOR register.

**Considerations:**

- Vector table is not necessarily at the start of the ram region: in the case that you also use code relocation, it will place the vector table after the relocated code. Relocated code will always be at the start of the RAM SECTION because of the hard coded generated value in file with "gen_relocate_app.py". It can create quite big gaps and lose about the size of alignment of the vector table section (which is 0x200 for siwx917). Either we consider to not have the vector table at the start of the SRAM section while using code_relocation (what is done is my code actually), or we need to modify the script that relocates the code.
- MPU is handled for the code relocation strategies and I think I need to handle this for the small ram section that I reserved for the Vector table. I have tried something concerning this consideration.
- VTOR register can be modified in HAL files so this modification can be discarded later in the code execution.
- VTOR can also be modified by MCU_BOOT with the software relay vector table and I did not handle that part for the moment.

Martinhoff-maker committed:
arch: arm: cortex_m: generalizing SRAM vector table relocation

Allow to place the vector table section in SRAM with
CONFIG_SRAM_VECTOR_TABLE option for all cortex-m architecture that
have VTOR register.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
tests: vector_table_relocation: add test for vector table relocation

Add a basic test for vector table relocation to SRAM. This test
verifies that the vector table can be relocated to SRAM and that
the VTOR register correctly points to the SRAM address range.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

andyross commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -39,22 +39,40 @@ Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used)) void *_vector_table
+
+#define VECTOR_ADDRESS ((uintptr_t)_sram_vector_start)
+
+static inline void relocate_vector_table(void)
```
Any particular reason you can't just leave this in .data and rely on the pre-existing copy code?

andyross commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -39,22 +39,40 @@ Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used)) void *_vector_table
+	(void) memcpy(_sram_vector_start, _vector_start, vector_size);
+	SCB->VTOR = VECTOR_ADDRESS & VTOR_MASK;
+	barrier_dsync_fence_full();
+	barrier_isync_fence_full();
```
The barrier instructions shouldn't be needed.   Interrupts don't get enabled until the switch to main, which involves a full context switch and implicit barriers from the PendSV.  And we absolutely shouldn't be tossing exceptions out of early init code. :)

andyross commented on arch/arm/core/cortex_m/ram_vector_table.ld:
```diff
@@ -0,0 +1,32 @@
+#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
+. = ALIGN( 1 << LOG2CEIL(4 * 64) );
+#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+. = ALIGN( 1 << LOG2CEIL(4 * 32) );
```
It's generally preferred to put the alignment onto the chunk in the object file instead of manually here.  Is that not already done?

andyross commented on arch/arm/core/mpu/arm_core_mpu.c:
```diff
@@ -106,6 +111,18 @@ static const struct z_arm_mpu_partition static_regions[] = {
+		.attr = K_MEM_PARTITION_P_R_U_RX,
+#else
+		.attr = K_MEM_PARTITION_P_RX_U_RX,
+#endif
```
This feels wrong.  You're burning a whole MPU register to support this feature when kernel .data is already walled off from userspace threads.

Martinhoff-maker replied:
I was thinking also to reuse an already configured MPU section like code relocation section (and it can also make sense to link CONFIG_CODE_DATA_RELOCATION_SRAM and CONFIG_SRAM_VECTOR_TABLE)

andyross reviewed (commented):
Notes and nitpicks.  Nothing fatal, feature is definitely useful.  Though ideally I guess we'd want to combine this with a "dynamic direct interrupts" feature with an app-visible API.  As it stands it requires an app to poke values directly into the table?

Martinhoff-maker commented:
Thanks for the early feedback @andyross. Yes combining with "dynamic direct interrupts" was in my mind since I'm going later (in another PR) to make use of "dynamic interrupt" to place sw_isr_table also in RAM. I'm currently trying to resolve the compilations problem with mps3 and will update the PR with your feedback.

jerome-pouiller commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -39,22 +39,40 @@ Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used)) void *_vector_table
+
+void __weak relocate_vector_table(void)
+{
+	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
```
I suggest an empty line between variables declarations and the code.

jerome-pouiller commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -39,22 +39,40 @@ Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used)) void *_vector_table
+void __weak relocate_vector_table(void)
+{
+	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
+	(void) memcpy(_sram_vector_start, _vector_start, vector_size);
```
Do you get a warning if you remove the cast?

Martinhoff-maker replied:
no, I can remove it.

jerome-pouiller commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -39,22 +39,40 @@ Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used)) void *_vector_table
 
+#ifdef CONFIG_SRAM_VECTOR_TABLE
+
+/* Vector table is in ROM, copy it to the RAM and set VTOR*/
```
Space before `*/` is missing.

jerome-pouiller commented on arch/arm/core/cortex_m/ram_vector_table.ld:
```diff
@@ -0,0 +1,32 @@
+ */
+
+/* Vector table is not necessarily at the start of the ram region: in the case that you also use
+ * code rellocation, it will place the vector table after the rellocated code. Rellocated code
```
"relocation"

jerome-pouiller commented on arch/arm/core/cortex_m/ram_vector_table.ld:
```diff
@@ -0,0 +1,32 @@
+
+SECTION_PROLOGUE(.sram_vt,,)
+{
+/* Heritage of vector table alignment in flash (see vector_table.ld in arch/arm/core/) */
```
I believe the code in `{ }` should be indented with one tab.

Martinhoff-maker replied:
yes

jerome-pouiller commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -39,22 +39,40 @@ Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used)) void *_vector_table
 
 #ifdef CONFIG_CPU_CORTEX_M_HAS_VTOR
```
Does `CONFIG_SRAM_VECTOR_TABLE` depend on `CONFIG_CPU_CORTEX_M_HAS_VTOR`? If yes, maybe the declaration of CONFIG_SRAM_VECTOR_TABLE should be updated.

Martinhoff-maker replied:
No CONFIG_SRAM_VECTOR_TABLE can be used when CONFIG_CPU_CORTEX_M_HAS_VTOR is not set but it is a special case with STM32f0 board

jerome-pouiller commented on arch/arm/core/mpu/arm_core_mpu.c:
```diff
@@ -106,6 +111,18 @@ static const struct z_arm_mpu_partition static_regions[] = {
+#if defined(CONFIG_ARM_MPU_PXN) && defined(CONFIG_USERSPACE)
+		.attr = K_MEM_PARTITION_P_R_U_RX,
+#else
+		.attr = K_MEM_PARTITION_P_RX_U_RX,
```
I believe `_sram_vector_start` only contains data. So, maybe `K_MEM_PARTITION_P_RO_U_RO` is sufficient? (unless it is strategy to coalesce the MPU segments)

ioannis-karachalios commented:
Hi @Martinhoff-maker, 
There is a recent [PR](https://github.com/zephyrproject-rtos/zephyr/pull/88588) that drops the usage of CONFIG_SRAM_VECTOR_TABLE for the Smartbond devices. In fact, the SoC supports a custom mechanism for remapping IVT to SYSRAM. VTOR is implemented in M33 but not used, at all.

Martinhoff-maker committed:
arch: add dependencie for CONFIG_SRAM_VECTOR_TABLE

Move SRAM_VECTOR_TABLE symbol from general Kconfig to Arch Kconfig
because it depends on the architecture possibility to relocate the
vector table.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

mathieuchopstm commented on arch/arm/core/cortex_m/ram_vector_table.ld:
```diff
@@ -0,0 +1,32 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Vector table is not necessarily at the start of the ram region: in the case that you also use
+ * code relocation, it will place the vector table after the relocated code. Relocated code
+ * will always be at the start of the RAM SECTION because of the hardcoded generated value in file
+ * with "gen_relocate_app.py". It can create quite big gap and lose about the size of alignment
+ * of RAM Space
```
Not sure I fully understand this. If code relocation is enabled, then the relocated vector table *could* be not placed at the start of SRAM?
Do we have no way to guarantee it will be at the start?

If we can't guarantee it, then ignore my other comment, but I think we ought to be able to do it...

Martinhoff-maker replied:
When code relocation is enabled, it will create a linker script file (linker_relocate.ld) where the start of the first relocation subsection will be at: ORIGIN(RAM) (if relocated to RAM MEMORY). So, this will override the placement mechanism with  linker script snippet. You can play with memory region (create a special memory region at the start for VT) to overpass this constraint, you can modify the script that generate the linker_relocate.ld ... it's a question of arch at this point I think. I do not see why we have such a big constraint like "ORIGIN(RAM)" in the linker_relocate.ld. IMO we should not have that constraint and moreover, code relocation mechanism should be compliant with the linker script snippet sorting mechanism.

mathieuchopstm commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -39,22 +39,41 @@ Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used)) void *_vector_table
+
+#define VECTOR_ADDRESS ((uintptr_t)_sram_vector_start)
+
+void __weak relocate_vector_table(void)
+{
+	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
+
+	memcpy(_sram_vector_start, _vector_start, vector_size);
+	SCB->VTOR = VECTOR_ADDRESS & VTOR_MASK;
+	barrier_dsync_fence_full();
+	barrier_isync_fence_full();
+}
```
Do we really have to duplicate this routine?
Why not just do the following in the existing implementation?
```suggestion
#ifdef CONFIG_SRAM_VECTOR_TABLE
#define VECTOR_ADDRESS ((uintptr_t)_sram_vector_start)
#else
#define VECTOR_ADDRESS ((uintptr_t)_vector_start)
#endif

void __weak relocate_vector_table(void)
{

#ifdef CONFIG_SRAM_VECTOR_TABLE
	/* Copy vector table to its location in SRAM */
	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;

	memcpy(_sram_vector_start, _vector_start, vector_size);
#endif
	SCB->VTOR = VECTOR_ADDRESS & VTOR_MASK;
	barrier_dsync_fence_full();
	barrier_isync_fence_full();
}
```
(note: had to put the diff here because GitHub won't allow `suggestion` on untouched code ranges)

mathieuchopstm commented on arch/Kconfig:
```diff
@@ -573,6 +573,13 @@ config IRQ_OFFLOAD
 
+config SRAM_VECTOR_TABLE
+	bool "Place the vector table in SRAM instead of flash"
+	depends on ARCH_HAS_VECTOR_TABLE_RELOCATION
```
Either here or at Cortex-M level:
```suggestion
	depends on ARCH_HAS_VECTOR_TABLE_RELOCATION || CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP
```
\+ see comment/proposal in `prep_c.c`

Martinhoff-maker replied:
More on cortex-M level IMO like this:
```
select ARCH_HAS_VECTOR_TABLE_RELOCATION if (CPU_CORTEX_M_HAS_VTOR || CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP)
```
But since, as you said, SRAM_VECTOR_TABLE symbol will be set to y as you redefined the symbol dependencies, it will not add anything (or maybe transparency for a user).

This change might go with the changes in prep.c in the cae we don't have a VTOR register.

mathieuchopstm replied:
> More on cortex-M level IMO like this:
> 
> ```
> select ARCH_HAS_VECTOR_TABLE_RELOCATION if (CPU_CORTEX_M_HAS_VTOR || CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP)
> ```

LGTM.
 
> But since, as you said, SRAM_VECTOR_TABLE symbol will be set to y as you redefined the symbol dependencies, it will not add anything (or maybe transparency for a user).

One reason I was proposing this is that I'd like the SoC to *not* need to redefine the symbol dependencies :slightly_smiling_face:

mathieuchopstm commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -39,22 +39,41 @@ Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used)) void *_vector_table
 
+#endif /* CONFIG_SRAM_VECTOR_TABLE */
 #else
 #define VECTOR_ADDRESS 0
 
```
On one hand, since `CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP` is architecture-wide, I would say we need to handle it here to avoid duplication. I think we could do something like this:
```suggestion
#ifdef CONFIG_SRAM_VECTOR_TABLE
#define VECTOR_ADDRESS ((uintptr_t)_sram_vector_start)
#else
/* Without VTOR, the vector table is mapped at address 0x0 */
#define VECTOR_ADDRESS (0x0)
#endif

void relocate_vector_table(void)
{
#if defined(CONFIG_XIP) && (CONFIG_FLASH_BASE_ADDRESS != 0) ||                                     \
	!defined(CONFIG_XIP) && (CONFIG_SRAM_BASE_ADDRESS != 0)
	/* Q: are we guaranteed to fall here if SRAM_VECTOR_TABLE=y?
	 * If no, we should add "|| defined(SRAM_VECTOR_TABLE)
	 */
	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
	(void)memcpy(VECTOR_ADDRESS, _vector_start, vector_size);
        z_arm_soc_map_vector_table(); //see below
#elif defined(CONFIG_SW_VECTOR_RELAY) || defined(CONFIG_SW_VECTOR_RELAY_CLIENT)
	_vector_table_pointer = _vector_start;
#endif
}

void __weak z_arm_soc_remap_vector_table() { /* Do nothing */ }
```

Then in `soc/st/stm32/stm32f0x/soc.c`, we would replace `void relocate_vector_table(void)` with:
```c
void z_arm_soc_remap_vector_table(void) {
    LL_SYSCFG_SetRemapMemory(
        COND_CODE_1(IS_ENABLED(CONFIG_SRAM_VECTOR_TABLE),
            (LL_SYSCFG_REMAP_SRAM),
            (LL_SYSCFG_REMAP_FLASH)));
}
```

---

On the other hand, the only SoC to have needed this so far is the STM32F0 series, so I'm questioning if correctness is worth all this trouble...

Martinhoff-maker replied:
I understand your point and think it can be a good idea ! Since it requires some code change on St soc file and the PR change does not break that part, I suggest to do this in another PR.
Concerning your question, We need to add " || CONFIG_SRAM_VECTOR_TABLE ".

mathieuchopstm replied:
Fine by me to postpone to another PR :slightly_smiling_face: 

Closing this point.

mathieuchopstm commented on tests/application_development/vector_table_relocation/src/main.c:
Maybe the bulk of this file should be moved to a Cortex-M specific file (e.g. `cortex_m.c`) to allow easily extending the test case for other architectures, if required in the future.

Martinhoff-maker replied:
Totally agree.

mathieuchopstm reviewed (commented):
Minor remark but otherwise LGTM.

Agree to postpone `CORTEX_M0_VECTOR_TABLE_REMAP` support in another PR since only STM32F0 uses the symbol, and this PR doesn't break it.

gramsay0 commented on arch/arm/core/cortex_m/CMakeLists.txt:
```diff
@@ -61,3 +61,17 @@ zephyr_linker_sources(
   )
 endif()
+
+if (CONFIG_SRAM_VECTOR_TABLE)
```
Yes, but that check is done twice:
```
zephyr_linker_sources_ifdef(CONFIG_SRAM_VECTOR_TABLE
```

MaureenHelm commented:
> 
> CONFIG_SRAM_VECTOR_TABLE is currently only used by some boards/soc:
> 
> * adi/max32 -> activate by default on all the max32 boards but no use of it as I see ? (might add MaureenHelm ADI maintener on the review)

That's correct, it's not being used by the max32 port. I think it was incorrectly copied from another port when the max32 port was first done.

jerome-pouiller commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -39,7 +39,11 @@ Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used)) void *_vector_table
+#define VECTOR_ADDRESS ((uintptr_t)_sram_vector_start)
+#else
 #define VECTOR_ADDRESS ((uintptr_t)_vector_start)
+#endif
```
Indeed, the Linux Coding Style says ["at the end of non-trivial #if or #ifdef block"][https://www.kernel.org/doc/html/v4.10/process/coding-style.html#conditional-compilation].

wearyzen commented on arch/arm/core/cortex_m/ram_vector_table.ld:
```diff
@@ -0,0 +1,32 @@
+
+	_sram_vector_start = .;
+	. += _vector_end - _vector_start;
+	MPU_ALIGN(_sram_vector_size);
```
maybe since MPU_ALIGN(_sram_vector_size); does the minimum alignment for MPU 🤷 which made it work
but I guess its alright to keep it consistent with the other scripts.

wearyzen commented on Kconfig.zephyr:
```diff
@@ -215,12 +215,6 @@ config LINKER_SORT_BY_ALIGNMENT
 	  in decreasing size of symbols. This helps to minimize
 	  padding between symbols.
 
-config SRAM_VECTOR_TABLE
```
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-4.2.rst#architectures looks like a good section to add this with something like "moved the config ... and added dependency to ... to support relocation of vector table in RAM"

wearyzen commented on tests/application_development/vector_table_relocation/prj.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_COVERAGE=n
+CONFIG_ZTEST=y
+CONFIG_SRAM_VECTOR_TABLE=y
```
Excluding mps3/corstone310 would be a good idea and atm using `depends on !ROMSTART_RELOCATION_ROM` should be ok. Both CONFIG_ROMSTART_RELOCATION_ROM and CONFIG_SRAM_VECTOR_TABLE are relatively uncommon, and enabling both simultaneously doesn’t appear to have a clear or valid use case. If a legitimate need arises in the future, we can revisit this.

JarmouniA reviewed (approved):
Looks good to me.

bjarki-andreasen reviewed (approved):
Looks good :)
